Add tunnel mode config and egress gateway config params

blzhao-0 opened the pull request:
This PR adds a tunnel mode arg to the init script of cilium devices so
that 'mode' and 'tunnel mode' is represented in separate params. It 
also adds the config options for enabling egress gateway and makes
sure that tunnel is enabled when egress gateway is specified.

This change is part the egress NAT feature, Ref: #13575.


Signed-off-by: Bolun Zhao <blzhao@google.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Thanks for contributing!

<!-- Description of change -->

blzhao-0 committed:
Refactor of tunnel mode and tunnel specific configs

This change adds a tunnel mode arg to the init script of cilium devices so
that 'mode' and 'tunnel mode' is represented in seperate params. The
functioning should remain the same.

This change is part of the ground work for the egress NAT feature, which
requires this tunnel config information Ref: #13575

Signed-off-by: Bolun Zhao <blzhao@google.com>

joestringer commented on bpf/init.sh:
```diff
@@ -481,10 +482,10 @@ else
 
-if [ "$MODE" = "vxlan" -o "$MODE" = "geneve" ]; then
-	ENCAP_DEV="cilium_${MODE}"
+if [ "$MODE" = "tunnel" ]; then
```
Could you double-check the rest of the `$MODE` usage in the file? It looks like there are places where it could be `ipip`, which I presume should be handled via the `${TUNNEL_MODE}` option as well.

anfernee replied:
Yes. `$MODE` could be `ipip`. It's not used for tunnel mode, but it should be covered. @MasterZ40 can you check it out and fix it?

blzhao-0 replied:
Thanks for the comments. 
Sorry if i did not fully understood, but it seems like `ipip` mode is handled separately in this [file.](https://github.com/cilium/cilium/blob/72949b4b49d43e6dbb4714929222acb688ce6dd2/bpf/init.sh#L441) @borkmann could you comment?

pchaigno replied:
@MasterZ40 The condition here should check for VXLAN or GENEVE. We don't want to create a `cilium_ipip` device for IPIP.

borkmann replied:
@MasterZ40 the ipip is handled separately, mostly given we need separate devices for v4/v6 which is not the case for vxlan + geneve. Also the kernel device name handling is more hacky for the ipip ones which ignores passed name when it needs to pull in the driver.

For testing ipip I typically use: `./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true --datapath-mode=lb-only --bpf-lb-algorithm=maglev --bpf-lb-maglev-table-size=2039 --bpf-lb-mode=dsr --bpf-lb-acceleration=native --devices=enp2s0np0 --bpf-lb-dsr-dispatch=ipip --disable-envoy-version-check=true`

blzhao-0 replied:
@borkmann thanks for the clarification.  @pchaigno I think in this case as Daniel said the ipip devices are handled separately as datapath tunnel devices. And this change is not really touching that separation. The ipip devices are handled here: https://github.com/cilium/cilium/blob/72949b4b49d43e6dbb4714929222acb688ce6dd2/bpf/init.sh#L441 

I might be wrong here, but let me know if this makes sense. Thx!

pchaigno replied:
Ah, I thought IPIP would be one of the tunnel modes, but it's actually one of the modes. Ok :+1:

joestringer reviewed (changes requested):
Hi @MasterZ40 , thanks for the PR.

The code change itself looks mostly straightforward, but I don't understand why. Would you mind providing a brief motivation in the commit message / PR description?

Please also review the rest of the PR description, everything below "Please ensure your pull request adheres to the following guidelines" is for you to complete or remove if it makes no sense. In particular the release note needs to be either set, or just removed (it's fine to remove in this case since the PR title describes the changes sufficiently).

One more specific change request below.

pchaigno commented on bpf/init.sh:
```diff
@@ -19,26 +19,27 @@ RUNDIR=$2
-NR_CPUS=${22}
-ENDPOINT_ROUTES=${23}
-PROXY_RULE=${24}
+# Set if MODE = "direct", "ipvlan", "flannel", "tunnel"
```
AFAIU, this comment is for `NATIVE_DEVS`, not for `TUNNEL_MODE`. We might need a similar comment for `TUNNEL_MODE` though.

blzhao-0 replied:
Fixed. Thanks for the feedback!

blzhao-0 committed:
Add option enable-egress-gateway

Adds "enable-egress-gateway" to the config, with defualt value to be
false.

Signed-off-by: Bolun Zhao <blzhao@google.com>

blzhao-0 committed:
Enable tunnel mode when egress gateway is configured

Sets the tunnel option when egress gateway is configured whilst direct
routing mode. Under this circumstance, the tunnel will be used to
redirect egress traffic to the gateway and it defaults to vxlan.

Signed-off-by: Bolun Zhao <blzhao@google.com>

joestringer reviewed (approved):
The changes themselves look fine to me. You'll need to take into account MTU for mixed mode as well, otherwise endpoints may end up transmitting packets that cannot fit in the tunnel (I believe this is currently switching on tunnel mode enablement).

Then there's just the question of whether to merge PRs that contain partially-finished features. There has been some ongoing debate about this. I think previously the PR was just refactoring so it's clear we would just merge it, but with the latest changes to also start including the beginnings of the feature for #13575, it's less clear.

EDIT: I guess on the latter aspect, the [ship has already departed](https://github.com/cilium/cilium/pull/14712). So I think the remaining item here is just the other codeowners.
